### PR TITLE
Snapshot editor

### PR DESCRIPTION
**Pull request by ShadowCurse**
## Changes
Added initial version of `snapshot-editor` tool.
Initial functionality is:
- (aarch64 only) remove registers from vcpu states
- (aarch64 only)  print all vcpu states from vmstate file

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: CHANGELOG.md
Comment (roypat) [inline CHANGELOG.md:8]: I think we can go a little bit more into details here. What exactly can it modify? Also, "snapshot files" is a bit ambiguous here, since it could refer to both the memory and the vmstate file. We should clarify that is only allows modification of the vmstate file.

Comment (roypat) [reply CHANGELOG.md:8]: Maybe it can also link to the documentation you add in this PR?

File: src/snapshot-editor/src/info.rs
Comment (roypat) [inline src/snapshot-editor/src/info.rs:40]: Mh, why does this closure take a `u16` for the version, while the one of the `edit` function in `edit_vmstate.rs` does not? Especially since the field is just ignored below

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (roypat) [inline tests/integration_tests/functional/test_snapshot_editor.py:72]: Maybe here we'll want to execute a command inside the VM to check everything works? e.g. a `new_vm.ssh.execute("true")` (e.g. whatever the other snapshot related tests do)

**Review by roypat — COMMENTED** [review]
Looks good to me, just the changelog could maybe explain a new tool a bit better

File: docs/snapshotting/snapshot-editor.md
Comment (roypat) [inline docs/snapshotting/snapshot-editor.md:13]: Right now it doesnt - we should write the documentation based on what currently is implemented (e.g. this tool deals with vmstate, and rebase-snap deals with memory file)

File: docs/snapshotting/snapshot-editor.md
Comment (roypat) [inline docs/snapshotting/snapshot-editor.md:61]: Maybe we can add some example output here?

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (roypat) [inline src/snapshot-editor/src/edit_vmstate.rs:75]: Maybe we can print a warning if some register's removal was requested, but the register isnt present?

File: docs/snapshotting/snapshot-editor.md
Comment (roypat) [inline docs/snapshotting/snapshot-editor.md:43]: Maybe this can briefly explain what the tool does? e.g. reads `vmstate_file`, removes the registers 0x1 and 0x2 and then writes the modified file to `new_vmstate_file`

**Review by roypat — APPROVED** [review]

File: src/snapshot-editor/src/info.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/info.rs]: initially the tool could also tell the snapshot/firecracker version, but I removed it to make PR smaller.
Removed the `u16` parameter.

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py:73]: Added a call to `ls` after restoration.

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/edit_vmstate.rs]: Added printing of which registers were removed and which regs were missing/not present.

File: docs/snapshotting/snapshot-editor.md
Comment (ShadowCurse) [reply docs/snapshotting/snapshot-editor.md:46]: There is a small description of the command. I think it is sufficient.

File: docs/snapshotting/snapshot-editor.md
Comment (roypat) [reply docs/snapshotting/snapshot-editor.md:13]: I still we should just say "`snapshot-editor` (currently) allows to modify the `vmstate` file of aarch64 snapshots" instead of having multiple notes. Because this sentence here might be true in the future, but currently it is still false, so we should just merge it with the notes.

Comment (ShadowCurse) [reply docs/snapshotting/snapshot-editor.md]: Merged

**Review by roypat — APPROVED** [review]

File: docs/snapshotting/snapshot-editor.md
Comment (roypat) [inline docs/snapshotting/snapshot-editor.md:13]: nit: 
```suggestion
`snapshot-editor` (currently) allows to modify only `vmstate` files only
```

File: src/snapshot-editor/src/main.rs
Comment (bchalios) [inline src/snapshot-editor/src/main.rs:9]: You can use [std::compile_error](https://doc.rust-lang.org/std/macro.compile_error.html) to avoid doing the `#[cfg(target_arch = "aarch64)]` thing multiple times. It is quite awkward otherwise that we produce a binary that does nothing on x86.

File: src/snapshot-editor/src/info.rs
Comment (bchalios) [inline src/snapshot-editor/src/info.rs:40]: Same as before, I think it's much simpler to just have a single function here that opens the file and prints the related info.

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (bchalios) [inline src/snapshot-editor/src/edit_vmstate.rs:62]: what's the point of passing this function pointer in `edit`? It looks to me it's a needless indirection.
Why not simply call `remove_regs` when you get the `EditVmStateSubCommand::RemoveRegs` subcommand?

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (bchalios) [reply src/snapshot-editor/src/edit_vmstate.rs:67]: Oh, is it meant to mention which vCPU you're currently editing? If so,

```
println!("Modifying state for vCPU {i}");
```

reads better IMHO.

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (bchalios) [inline src/snapshot-editor/src/edit_vmstate.rs:24]: It'd be nice to mention what these `u64` values actually are. Are they indexes, some other name, or key?

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (bchalios) [inline src/snapshot-editor/src/edit_vmstate.rs:80]: This seems a bit difficult to read. Would something like that:

```rust
for reg in vcpu_state.regs.iter() {
   if !remove_regs.contains(&reg.id) {
      new_regs.push(reg);
   } else {
}
```

not work?

Comment (bchalios) [reply src/snapshot-editor/src/edit_vmstate.rs:80]: oh sorry, I see you do that to see if the user requested to delete a register that is not present. Do we really care about that?

File: docs/snapshotting/snapshot-editor.md
Comment (bchalios) [inline docs/snapshotting/snapshot-editor.md:31]: Here, as well, we should describe what type of things are `REGS`

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (bchalios) [inline tests/integration_tests/functional/test_snapshot_editor.py:68]: I think this test can be more generic. I can think of at least two cases to test here:

1. We should check that no other registers were touched or added.
2. We should check that if we ask the tool to remove a non-existent register it does the right thing.

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (bchalios) [inline tests/integration_tests/functional/test_snapshot_editor.py:78]: I guess these need to be adapted to the new style of checking errors, right?

File: src/snapshot/src/lib.rs
Comment (ShadowCurse) [reply src/snapshot/src/lib.rs:144]: it is part of the 'snapshot'. The `O::deserialize` in this case can be thought as `MicrovmState::deserialize`. And `MicrovmState` does not have version inside of it.

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (ShadowCurse) [reply src/vmm/src/vstate/vcpu/aarch64.rs:256]: Actually opposite. I prefer to have fields to be public. The prevention of changes to these fields should be with usage of `const` or `mut` modifiers.

File: src/snapshot-editor/src/main.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs:43]: On x86_64 the tool actually has not functionality, so `#[allow(unused)]` is needed to allow `cli` variable to be unused. The follow up PRs should change this.

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/edit_vmstate.rs:62]: The idea here is to have a common code for all `edit` methods in one method. So if we add new `edit` methods it can be used as such:
```rust
 match command {
        SubCommand::RemoveRegs => edit(&vmstate_path, &output_path, |state| {
            remove_regs(state, &regs)
        })?,
        StateSubCommand::SomeEditMethod2 => edit(&vmstate_path, &output_path, some_edit_method_2)?,
        StateSubCommand::SomeEditMethod3 => edit(&vmstate_path, &output_path, some_edit_method_3)?,
}
```

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (bchalios) [reply src/vmm/src/vstate/vcpu/aarch64.rs:256]: It's not about preventing modification of the field, it's about being able to change the implementation without breaking the anybody who consumes the API, e.g. if we ever change `old_regs` to be stored in different kind of container you can still have a getter that returns a `Vec<Aarch64RegisterOld>` so that anybody calling that function won't break.

File: src/snapshot-editor/src/main.rs
Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: yup, exactly. And then you can get rid of all the other directives (target_arch and unused)

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (bchalios) [reply src/snapshot-editor/src/edit_vmstate.rs:62]: I got it, I just think it's unnecessary indirection. What's the benefit of this over:

```rust
 match command {
        SubCommand::RemoveRegs => remove_regs(&vmstate_path, &output_path)?,
        StateSubCommand::SomeEditMethod2 => some_edit_method_2(&vmstate_path, &output_path)?,
        StateSubCommand::SomeEditMethod3 => some_edit_method_3(&vmstate_path, &output_path)?,
}
```

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (ShadowCurse) [reply src/vmm/src/vstate/vcpu/aarch64.rs:256]: Considering we are the only ones consuming our internal API, we can handle the changes. Getters/setters just increase indirection and complexity and even in our code we don't use them much.

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/edit_vmstate.rs:62]: I guess it is just personal preference. Otherwise each edit method will start with ` let (microvm_state, version) = open_vmstate(vmstate_path)?;` and end with `save_vmstate(microvm_state, output_path, version)?;`. Considering how small the amount of code is, both versions can be correct.

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (pb8o) [inline tests/integration_tests/functional/test_snapshot_editor.py:1]: 2023 :)

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (pb8o) [inline tests/integration_tests/functional/test_snapshot_editor.py:44]: Simpler: `assert MIDR_EL1 in stdout`

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (pb8o) [inline tests/integration_tests/functional/test_snapshot_editor.py:20]: Is there any value in testing with all guest kernels?

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (pb8o) [inline tests/integration_tests/functional/test_snapshot_editor.py:68]: Simpler: `assert MIDR_EL1 not in stdout`

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (bchalios) [reply src/vmm/src/vstate/vcpu/aarch64.rs:256]: Everybody can handle the changes. I'm just suggesting that we do it with the minimum amount of effort. While at the same time avoiding making public things that otherwise shouldn't be.

File: src/snapshot-editor/src/main.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs:23]: fixed

File: src/snapshot-editor/src/main.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs:9]: wait, this will not work, because when we build with `./tool/devtool build` it build everything, including `snapshot-editor`. And on x86 this build will fail. And CI will fail.

File: src/snapshot-editor/src/info.rs
Comment (bchalios) [reply src/snapshot-editor/src/info.rs:40]: Ok, the previous case is fine. But here, will we ever do something with print info command other than printing info?

Comment (ShadowCurse) [reply src/snapshot-editor/src/info.rs:40]: hmm... printing more info? I thought of adding print command for each data type in vmstate: vm memory, device states... etc.

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: isn't uvm_nano 5.10 only?

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py]: fixed

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py]: fixed

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py]: fixed

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py:78]: what is new way?

File: docs/snapshotting/snapshot-editor.md
Comment (ShadowCurse) [reply docs/snapshotting/snapshot-editor.md:38]: Added parameters descriptions

File: src/snapshot-editor/src/edit_vmstate.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/edit_vmstate.rs]: Added another comment

File: src/snapshot-editor/src/main.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs]: fixed

File: docs/snapshotting/snapshot-editor.md
Comment (ShadowCurse) [reply docs/snapshotting/snapshot-editor.md]: fixed

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_editor.py:78]: oh sorry forget that. I thought this was an API call that would trigger an exception.

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (pb8o) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: Yeah, that's why I am asking if running in 4.14 could have different behavior.

File: src/snapshot-editor/src/main.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs:9]: I don't think this is worth it. There will be a follow up PR with additional functionality available on both x86 ant aarch64, so x86 version will be not useful only for short period of time.

File: docs/snapshotting/snapshot-editor.md
Comment (ShadowCurse) [reply docs/snapshotting/snapshot-editor.md]: fixed

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: The test itself won't change behavior. Set of registers in the snapshot is same for all kernels.

File: src/snapshot-editor/src/main.rs
Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: It doesn't look too complicated to me:

```diff
diff --git a/tools/release.sh b/tools/release.sh
index 1c84e510..2fb55039 100755
--- a/tools/release.sh
+++ b/tools/release.sh
@@ -60,6 +60,7 @@ function get-firecracker-version {
 # defaults
 LIBC=musl
 PROFILE=dev
+ARCH=$(uname -m)
 MAKE_RELEASE=

 #### Option parsing
@@ -130,7 +131,12 @@ fi

 say "Building version=$VERSION, profile=$PROFILE, target=$CARGO_TARGET..."
 # shellcheck disable=SC2086
-cargo build --target "$CARGO_TARGET" $CARGO_OPTS --workspace
+if [[ $ARCH != "aarch64" ]]; then
+    cargo build --target "$CARGO_TARGET" $CARGO_OPTS --workspace --exclude snapshot-editor
+else
+    cargo build --target "$CARGO_TARGET" $CARGO_OPTS --workspace
+fi
+

 say "Binaries placed under $CARGO_TARGET_DIR"
```

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: what about 6.1?

File: src/snapshot-editor/src/main.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs:9]: The idea here is that this change will be reverted when x86 functionality will be added. Most `#[cfg(target_arch = "aarch64")]` will be removed anyway.

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: Same

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (roypat) [reply src/vmm/src/vstate/vcpu/aarch64.rs:256]: I think in Rust, generally direct field access is preferred over getters/setters, mainly because Rust's ownership rules mean that there is multiple ways of "getting" data (e.g. a borrow vs moving), and also because direct field access allows for pattern matching, which is more idiomatic than a sequence of `if`s on `.get_(...)` calls.

File: docs/snapshotting/snapshot-editor.md
Comment (roypat) [reply docs/snapshotting/snapshot-editor.md]: _technically_, its "the kernel virtual machine" xP English is weird

File: src/snapshot-editor/src/main.rs
Comment (roypat) [reply src/snapshot-editor/src/main.rs:9]: I think the change in devtool makes sense, because why would we ship a binary to our customers that does nothing. It would just be confusing. If we do the devtool change, I dont think we need to change anything about the conditional compilation: We dont ship it, so it doesnt matter if it builds and produces a binary that does nothing, or if it doesnt even build.

Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: I know. I 'd just like to have a way to explicitly not build the tool for x86 until it actually does something on x86.

Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs:9]: Ok I added changes to the `release.sh`

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (bchalios) [reply src/vmm/src/vstate/vcpu/aarch64.rs:256]: Ok, we're going into philosophical questions now :P AFAIK, for the time being, Rust seems to have issues with getting/moving struct fields comparing to the whole struct, but in any case, my main issue is about hiding implementation from the API, not ownership.

In any case, this is not a roadblock for me (marked as nit). Feel free to resolve Egor.

File: tests/integration_tests/functional/test_snapshot_editor.py
Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: my point here is that, we cannot be sure that it will always be the same, so maybe it makes sense to add tests for the rest of the kernels as well? WDYT @pb8o @ShadowCurse ?

Comment (ShadowCurse) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: As I mentioned before, this test will behave same for any kernel, because MIDR_EL1 reg will be in the snapshot in any case. If we want to have a test, that verifies that the set of regs is the same on all kernels, it should be a separate test.

Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_editor.py:20]: You are right. That's fine by me then.

File: src/snapshot-editor/src/main.rs
Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs]: Well.. CI still fails because it still tries to build x86 version of `snapshot-editor`. I would prefer to have
 `#[cfg(target_arch = "aarch64")]` rather then `std::compile_error!()` in this case.

Comment (ShadowCurse) [reply src/snapshot-editor/src/main.rs:9]: Reverted some changes. Left `#[cfg(target_arch = "aarch64")]` only in `main.rs`

Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: I'd still like to not ship a binary that does something in the release for x86 (this implies fixing the tests), but I'd like to hear as well from others, @pb8o, @roypat

Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: This patch should fix the CI, btw: 

```diff
diff --git a/tests/host_tools/cargo_build.py b/tests/host_tools/cargo_build.py
index dffe0479..67af0e00 100644
--- a/tests/host_tools/cargo_build.py
+++ b/tests/host_tools/cargo_build.py
@@ -63,6 +63,8 @@ def cargo_test(path, extra_args=""):
         "RUST_BACKTRACE": 1,
         "RUSTFLAGS": get_rustflags(),
     }
+    if platform.machine() != "aarch64":
+        extra_args += "--exclude snapshot-editor"
     cargo("test", extra_args + " --all --no-fail-fast", env=env)


diff --git a/tests/integration_tests/build/test_clippy.py b/tests/integration_tests/build/test_clippy.py
index dc4ae415..7dafd797 100644
--- a/tests/integration_tests/build/test_clippy.py
+++ b/tests/integration_tests/build/test_clippy.py
@@ -22,4 +22,5 @@ def test_rust_clippy(target):
     """
     Test that clippy does not generate any errors/warnings.
     """
-    cargo("clippy", f"--target {target} --all --profile test", "-D warnings")
+    extra_args = "--exclude snapshot-editor" if platform.machine() != "aarch64" else ""
+    cargo("clippy", f"--target {target} --all --profile test {extra_args}", "-D warnings")
diff --git a/tests/integration_tests/build/test_coverage.py b/tests/integration_tests/build/test_coverage.py
index ac3af0b2..96714159 100644
--- a/tests/integration_tests/build/test_coverage.py
+++ b/tests/integration_tests/build/test_coverage.py
@@ -42,7 +42,7 @@ def test_coverage(monkeypatch):
     # Generate test profiles.
     cargo(
         "test",
-        f"--all --target {TARGET}",
+        f"--all --target {TARGET} --exclude snapshot-editor",
         "--test-threads=1",
         env={
             "RUSTFLAGS": "-Cinstrument-coverage",
diff --git a/tests/integration_tests/build/test_unittests.py b/tests/integration_tests/build/test_unittests.py
index 8bcb80cb..79f42bcd 100644
--- a/tests/integration_tests/build/test_unittests.py
+++ b/tests/integration_tests/build/test_unittests.py
@@ -21,10 +21,10 @@ def test_unittests(test_fc_session_root_path):
     Run unit and doc tests for all supported targets.
     """
     extra_args = "--target {} ".format(TARGET)
-
     host.cargo_test(test_fc_session_root_path, extra_args=extra_args)


 def test_benchmarks_compile():
     """Checks that all benchmarks compile"""
-    host.cargo("bench", f"--all --no-run --target {TARGET}")
+    extra_args = "--exclude snapshot-editor" if platform.machine() != "aarch64" else ""
+    host.cargo("bench", f"--all --no-run --target {TARGET} {extra_args}")
```

Comment (roypat) [reply src/snapshot-editor/src/main.rs:9]: But I thought with the patch to release.sh, we are no longer shipping a binary for x86_64 as part of the release?

as an aside, i think the patch for the CI can do directly into `def cargo(...)` in host_tools.py, that'd make for a smaller change to later revert.

Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: true, we can keep the `release.sh` changes and we're fine with the release binary. But we ship source code as well, so someone could get the code, build the tool and wait for it to do something on x86. It just sounds weird (apart from it being ugly). The CI changes are small enough (good catch Patrick about the `host.cargo` command) to just do them and then revert a single line once we add x86 functionality.

Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: If everybody else is fine with it, I won't block the PR for this.

Comment (bchalios) [reply src/snapshot-editor/src/main.rs:9]: @roypat @pb8o could you chime in? this is the last open blocker

**Review by roypat — DISMISSED** [review]

**Review by bchalios — APPROVED** [review]

**Review by roypat — APPROVED** [review]

**Pull request merged**
